builtin/providers/cloudstack/resource_cloudstack_instance.go

```diff
@@ -186,6 +195,7 @@ func resourceCloudStackInstanceRead(d *schema.ResourceData, meta interface{}) er
 	d.Set("display_name", vm.Displayname)
 	d.Set("ipaddress", vm.Nic[0].Ipaddress)
 	d.Set("zone", vm.Zonename)
+	//NB cloudstack sometimes sends back the wrong keypair name, so dont update it
```

Comment:
Wow, that sounds like a nasty bug! Let me check this one with some CloudStack guys I work with...